cmake_files_header.md

```diff
@@ -0,0 +1,131 @@
+# CMake flags
+
+## Developer's guide for adding new CMake options
+
+### Don't be obvious. Be informative.
+
+Bad:
+```cmake
+option (ENABLE_TESTS "Enables testing" OFF)
+```
+
+This description is quite useless as is neither gives the viewer any additional information nor explains the option
+purpose. If the option's name is quite self-descriptive, prefer the empty description.
+
+Better:
+
+```cmake
+option(ENABLE_TESTS OFF)
+```
+
+If the option's purpose can't be guessed by its name, or the purpose guess may be misleading, leave a comment above
+the `option()` line and explain what it does. The best way would be linking the docs page (if it exists).
+The comment is parsed into a separate column (see below).
+
+Even better (default off value is omitted):
+
+```cmake
+# Adds the ability to test ClickHouse using Google.Test (would produce another target unit_tests_dbms).
+# see tests/CMakeLists.txt for implementation detail.
+option(ENABLE_GTEST_TESTS)
+```
+
+### If the option's state could produce unwanted (or unusual) result, explicitly warn the user.
+
+Suppose you have an option that may strip debug symbols from the ClickHouse's part.
+This can speed up the linking process, but produces a binary that cannot be debugged.
+In that case, prefer explicitly raising a warning telling the developer that he may be doing something wrong. 
+Also, such options should be disabled if applies.
+
+Bad:
+```cmake
+option(STRIP_DEBUG_SYMBOLS_FUNCTIONS
+    "Do not generate debugger info for ClickHouse functions.
+    ${STRIP_DSF_DEFAULT})
+
+if (STRIP_DEBUG_SYMBOLS_FUNCTIONS)
+    target_compile_options(clickhouse_functions PRIVATE "-g0")
+endif()
+
+```
+Better:
+
+```cmake
+# Provides faster linking and lower binary size.
+# Tradeoff is the inability to debug some source files with e.g. gdb
+# (empty stack frames and no local variables)."
+option(STRIP_DEBUG_SYMBOLS_FUNCTIONS
+    "Do not generate debugger info for ClickHouse functions."
+    ${STRIP_DSF_DEFAULT})
+
+if (STRIP_DEBUG_SYMBOLS_FUNCTIONS)
+    message(WARNING "Not generating debugger info for ClickHouse functions")
+    target_compile_options(clickhouse_functions PRIVATE "-g0")
+endif()
+```
+
+### In the option's description, explain WHAT the option does rather than WHY it does something.
+
+The WHY explanation should be placed in the comment.
+You may find that the option's name is self-descriptive.
+
+Bad:
```

Comment:
Cannot agree. It's Ok to make as detailed description as possible. It does not make a big difference to move some part out. Some people are using GUI tools (like `ccmake`) and they should see the full info.

Reply:
The thing is, ClickHouse has no developer docs, so to keep this consistent, we should better link the official docs site.